src/mapgen/mapgen.cpp

```diff
@@ -437,42 +437,58 @@ void Mapgen::setLighting(u8 light, v3s16 nmin, v3s16 nmax)
 
 void Mapgen::lightSpread(VoxelArea &a, v3s16 p, u8 light)
 {
-	if (light <= 1 || !a.contains(p))
-		return;
+	std::deque<std::pair<v3s16, u8>> queue;
+	queue.emplace_back(std::make_pair(p, light));
 
-	u32 vi = vm->m_area.index(p);
-	MapNode &n = vm->m_data[vi];
+	/*
+		The loop contains a recursive algorithm, but the wrapping
+		code runs it iteratively so we don't hit stack limits.
+	*/
+	while (!queue.empty()) {
+		{
+			auto &item = queue.front();
+			p = item.first;
+			light = item.second;
+		}
+		queue.pop_front();
 
-	// Decay light in each of the banks separately
-	u8 light_day = light & 0x0F;
-	if (light_day > 0)
-		light_day -= 0x01;
+		if (light <= 1 || !a.contains(p))
```

Comment:
`n.param1` also needs to be updated for `light = 1`, but spreading code can be skipped.
IMO comparing this nibble-split byte by non-zero values can be dangerous to miss edge cases where the upper/lower nibbles differ.